presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsAnnotationProcessor.java

```diff
@@ -0,0 +1,243 @@
+                    statValue = sourceStats.get(sourceStatsArgumentIndex);
+                    break;
+                case ROW_COUNT:
+                    statValue = outputRowCount;
+                    break;
+                case NON_NULL_ROW_COUNT:
+                    statValue = outputRowCount * (1 - firstFiniteValue(nullFraction, 0.0));
+                    break;
+                case USE_TYPE_WIDTH_VARCHAR:
+                    statValue = returnNaNIfTypeWidthUnknown(getTypeWidthVarchar(callExpression.getArguments().get(sourceStatsArgumentIndex).getType()));
+                    break;
+            }
+        }
+        return statValue;
+    }
+
+    private static int getTypeWidthVarchar(Type argumentType)
+    {
+        if (argumentType instanceof VarcharType) {
+            if (!((VarcharType) argumentType).isUnbounded()) {
+                return ((VarcharType) argumentType).getLengthSafe();
+            }
+        }
+        return -VarcharType.MAX_LENGTH;
+    }
+
+    private static double returnNaNIfTypeWidthUnknown(int typeWidthValue)
+    {
+        if (typeWidthValue <= 0) {
+            return NaN;
+        }
+        return typeWidthValue;
+    }
+
+    private static int getReturnTypeWidth(CallExpression callExpression, StatsPropagationBehavior operation)
+    {
+        if (callExpression.getType() instanceof FixedWidthType) {
+            return ((FixedWidthType) callExpression.getType()).getFixedSize();
+        }
+        if (callExpression.getType() instanceof VarcharType) {
+            VarcharType returnType = (VarcharType) callExpression.getType();
+            if (!returnType.isUnbounded()) {
+                return returnType.getLengthSafe();
+            }
+            if (operation == SUM_ARGUMENTS || operation == SUM_ARGUMENTS_UPPER_BOUNDED_TO_ROW_COUNT) {
+                // since return type is an unbounded varchar and operation is SUM_ARGUMENTS,
+                // calculating the type width by doing a SUM of each argument's varchar type bounds - if available.
+                int sum = 0;
+                for (RowExpression r : callExpression.getArguments()) {
+                    int typeWidth;
+                    if (r instanceof CallExpression) { // argument is another function call
+                        typeWidth = getReturnTypeWidth((CallExpression) r, UNKNOWN);
+                    }
+                    else {
+                        typeWidth = getTypeWidthVarchar(r.getType());
+                    }
+                    if (typeWidth < 0) {
+                        return -VarcharType.MAX_LENGTH;
+                    }
+                    sum += typeWidth;
+                }
+                return sum;
+            }
+        }
+        return -VarcharType.MAX_LENGTH;
+    }
+
+    // Return first 'finite' value from values, else return values[0]
+    private static double firstFiniteValue(double... values)
+    {
+        checkArgument(values.length > 1);
```

Comment:
> 0?

That is, a single element list is OK? or just check values is not empty

Reply:
We expect at least two arguments in this private function.